walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (request.getAttendees().isEmpty()) {
+      return free_time;
+    }
+
+    /**
+     * Go through every event. If an event has attendees of the requested meeting, then check if the
+     * time overlaps with our free time. If yes then remove that conflicted time slot from our free
+     * time.
+     */
+    for (Event event : events) {
+      if (!Collections.disjoint(new ArrayList<>(request.getAttendees()), event.getAttendees())) {
+        int conflicted_slot_index = containsEvent(free_time, event.getWhen());
+        if (conflicted_slot_index != -1) {
+          while(conflicted_slot_index < free_time.size()) {
+            //loop through free time slots till the event ends and remove it from our free time.
+            TimeRange free_slot = free_time.get(conflicted_slot_index);
+            free_time =
+                getNewFreeTime(free_slot, event.getWhen(), free_time, conflicted_slot_index);
+            conflicted_slot_index++;
+            //break if the event ends in one of the free slots
+            if (event.getWhen().end() < free_slot.end()) break;
+          }
+        }
+      }
+    }
+
+    // If any of the TimeRange in our freetime list has a
+    // duration less than the requested meeting duration then remove that slot.
+    free_time = checkDuration(free_time, meeting_duration);
+
+    return free_time;
+  }
+
+  // This method checks if an event overlaps with any of our free time slots.
+  // It returns the conflicted slot's
+  private int containsEvent(ArrayList<TimeRange> free_time, TimeRange event_timerange) {
+    for (TimeRange slot : free_time) {
+      if (slot.overlaps(event_timerange)) return free_time.indexOf(slot);
+    }
+    return -1;
+  }
+
+  /**
+   * This method free time , conflicted free time slot and current event which we are looping
+   * through and checks all conditions in which they could overlap. Then returns the free time minus
+   * overlapping part.
+   */
+  private ArrayList<TimeRange> getNewFreeTime(
+      TimeRange free_slot,
+      TimeRange event_slot,
+      ArrayList<TimeRange> free_time,
+      int conflicted_slot_index) {
+
+    int free_slot_start = free_slot.start();
+    int free_slot_end = free_slot.end();
+
+    int event_slot_start = event_slot.start();
+    int event_slot_end = event_slot.end();
+
+    // Case: Free time: |---|
+    // Event slot:    |-------|
+    if (event_slot_start < free_slot_start && event_slot_end > free_slot_end) {
+      free_time.remove(free_slot);
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:     |-----|
+    // or
+    // |------|
+    // |--|
+    else if (event_slot_start <= free_slot_start && event_slot_end < free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(TimeRange.fromStartEnd(event_slot_end, free_slot_end, false)));
```

Comment:
Could you use `.set` to update the element at the index position? Same below when replacing time slots.

Reply:
Thanks for letting me know will do that.